grpc/stream.go

```diff
@@ -0,0 +1,325 @@
+package grpc
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/dop251/goja"
+	"github.com/grafana/xk6-grpc/lib/netext/grpcext"
+	"github.com/mstoykov/k6-taskqueue-lib/taskqueue"
+	"github.com/sirupsen/logrus"
+	"go.k6.io/k6/js/common"
+	"go.k6.io/k6/js/modules"
+	"go.k6.io/k6/metrics"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+type message struct {
+	msg []byte
+}
+
+type stream struct {
+	vu     modules.VU
+	client *Client
+
+	logger logrus.FieldLogger
+
+	methodDescriptor protoreflect.MethodDescriptor
+
+	method string
+	stream *grpcext.Stream
+
+	tagsAndMeta    *metrics.TagsAndMeta
+	tq             *taskqueue.TaskQueue
+	builtinMetrics *metrics.BuiltinMetrics
+	obj            *goja.Object // the object that is given to js to interact with the stream
+
+	closeSend chan struct{}
+	done      chan struct{}
+
+	writeQueueCh chan message
+
+	eventListeners *eventListeners
+}
+
+// defineStream defines the goja.Object that is given to js to interact with the Stream
+func defineStream(rt *goja.Runtime, s *stream) {
+	must(rt, s.obj.DefineDataProperty(
+		"on", rt.ToValue(s.on), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+
+	must(rt, s.obj.DefineDataProperty(
+		"write", rt.ToValue(s.write), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+
+	must(rt, s.obj.DefineDataProperty(
+		"end", rt.ToValue(s.end), goja.FLAG_FALSE, goja.FLAG_FALSE, goja.FLAG_TRUE))
+}
+
+func (s *stream) beginStream() error {
+	req := &grpcext.StreamRequest{
+		Method:           s.method,
+		MethodDescriptor: s.methodDescriptor,
+	}
+
+	stream, err := s.client.conn.NewStream(s.vu.Context(), *req, metadata.MD{})
+	if err != nil {
+		return fmt.Errorf("failed to create a new stream: %w", err)
+	}
+	s.stream = stream
+
+	go s.loop()
+
+	return nil
+}
+
+func (s *stream) loop() {
+	ctx := s.vu.Context()
+	wg := new(sync.WaitGroup)
+
+	defer func() {
+		wg.Wait()
+		s.tq.Close()
+	}()
+
+	// read & write data from/to the stream
+	wg.Add(2)
```

Comment:
I would prefer to have Add in the same place where we do the Done so it is easier to follow the path.

Reply:
>I would prefer to have Add in the same place where we do the Done so it is easier to follow the path.

Could you please elaborate here? The `Add(2)` here adds expectations about following go routines. Do you want something like:

```golang
wg.Add(1)
go s.readData(wg)
wg.Add(1)
go s.writeData(wg)
```

or what?

Reply:
Ah, you meant to move them inside :facepalm:

Reply:
Yeah, I see you did already for another one.

Reply:
Resolved with the latest push

Reply:
Unfortunately this technically introduces a race. 

As the goroutine can not start for a while there is every chance that this function can go to the select loop, the loop can return, to go to `wg.Wait` and for the goroutine to still have not gotten to the `wg.Add` part.

I don't particularly like the fact that we are giving the wait group around as well, but do not have better solution then that as the usual way of 
```go
wg.Add(1)
go func() {
  defer wg.Done()
  readPump()
}()
```

doesn't work for one of the cases where we spawn one more goroutine inside.

While I doubt we can hit this race and from a curtesy look at the code there will not really be any problems, I still prefer if we revert this change.